### DISCUSSION TITLE
SEB crashes after login with Azure AD multi factor authentication in Moodle

### DISCUSSION THREAD

**Discussion by AntonT76** [SafeExamBrowser/seb-win-refactoring#478] [Q&A]
Hello,

we installed the new version of SEB (3.4.0). We are using Moodle (3.9.*) with Open ID Connect authentication via Microsoft Azure AD and multi factor authentication. After pressing the login button in Moodle and Moodle will forward the user to the MF authentication site every times SEB will crash. Attached you can find a screenshot and SEB log files.

We changed a lot of SEB settings (eg: deactivate URL filtering, activate session handling) but it did not help.

Exception:
- login with older version SEB 3.2.0 works
- if MF authentication will be done before using SEB it will work (-> MF authentication in SEB isn't required)

How can we solve this problem?

thanks, br Anton
![SEB-error](https://user-images.githubusercontent.com/66412602/193740017-39fe6dae-8627-48e3-aa52-06f1b324faa1.png)
[2022-10-03_15h53m48s_Browser.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9702963/2022-10-03_15h53m48s_Browser.log)
[2022-10-03_15h53m48s_Client.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9702964/2022-10-03_15h53m48s_Client.log)
[2022-10-03_15h53m48s_Runtime.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9702965/2022-10-03_15h53m48s_Runtime.log)

**Comment by dbuechel** [comment]
It would be great if you could provide the crash data saved in the system logs: They can be found in the Windows Event Viewer ("Ereignisanzeige" in German) under "Windows Logs" > "Application" (resp. "Anwendung" in German). There should be one or more error events at around 15:53 on date 2022-10-03. Given the information in the SEB logs, it does appear that there is a bug in the accessibility features we added in version 3.4.0, but it would be great if we had the system log data to verify that assumption.

@yolpsoftware Could you please have a look at the issue, it seems that the JavaScript calls made in the browser window fail and then lead to a crash of the browser engine: https://github.com/SafeExamBrowser/seb-win-refactoring/blob/master/SafeExamBrowser.UserInterface.Desktop/Windows/BrowserWindow.xaml.cs#L537.

**Reply by yolpsoftware** [reply]
@dbuechel not sure how I can debug this. Wouldn't a Javascript error just result in an error log? Under what circumstances can a Javascript error bring down the entire browser engine?

**Reply by dbuechel** [reply]
> Not sure how I can debug this

You could try to log in to an account requiring 2FA, or maybe @AntonT76 could provide you with a test configuration / login?

> Wouldn't a Javascript error just result in an error log?

One would assume so, but unfortunately not. It depends (at least with the engine we're using) on where, how and when JavaScript execution is attempted.

> Under what circumstances can a Javascript error bring down the entire browser engine?

When e.g. attempting to execute JavaScript before the DOM is completely loaded or the V8 context (JavaScript runtime) is initialized. I've already fixed such a crash which our testing coincidentally discovered (see https://github.com/SafeExamBrowser/seb-win-refactoring/commit/9d7b89d36c1561cc87feb8be58ae13f3d896a0c3). SEB would in that case crash (resp. the browser engine, which in turn makes the client component crash), but only on slow / virtual machines. As far as I can determine, this happened exactly due to the fact that on slow machines, the browser engine wasn't yet ready to execute JavaScript.

**Reply by AntonT76** [reply]
@yolpsoftware , @dbuechel - regarding the test account: I have to ask our IT services if it is possible to get an "official" account with 2FA. I will inform you if I have more information.
br, Anton

**Reply by dbuechel** [reply]
According to the event log entry, there was a load error and the crash happened in the end due to an unhandled exception in the load error event handler:

```
System.Exception
   at CefSharp.Internals.CefFrameWrapper.ThrowIfFrameInvalid()
   at CefSharp.Internals.CefFrameWrapper.get_IsMain()
   at SafeExamBrowser.Browser.BrowserControl.<Initialize>b__31_12(System.Object, CefSharp.LoadErrorEventArgs)
   at CefSharp.WinForms.ChromiumWebBrowser.CefSharp.Internals.IWebBrowserInternal.OnLoadError(CefSharp.LoadErrorEventArgs)
   at <Module>.CefSharp.Internals.ClientAdapter.OnLoadError(CefSharp.Internals.ClientAdapter*, scoped_refptr<CefBrowser>*, scoped_refptr<CefFrame>*, cef_errorcode_t, CefStringBase<CefStringTraitsUTF16>*, CefStringBase<CefStringTraitsUTF16>*)
```

So it would definitely be great if we had a testing account (or it might even suffice to have the URL to the 2FA authentication page) so that we can debug and test a possible fix. I'm now unsure whether it indeed has anything to do with the JavaScript execution or if it is a simple missing check when accessing the frame in the event handler: https://github.com/SafeExamBrowser/seb-win-refactoring/blob/master/SafeExamBrowser.Browser/BrowserControl.cs#L110.

**Reply by yolpsoftware** [reply]
@dbuechel line 105 of the Client log is enough to diagnose the problem. See my proposed fix below.

**Comment by AntonT76** [comment]
Dear @dbuechel 

Due to upcoming exams, we have switched back to the old version 3.2 for most of our examination rooms.
Today I did another test with version 3.4 and attached you can find the log files and the event viewer files. I hope these files can help to solve the problem.

br, Anton
[2022-10-05_09h04m27s_Browser.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9713614/2022-10-05_09h04m27s_Browser.log)
[2022-10-05_09h04m27s_Client.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9713615/2022-10-05_09h04m27s_Client.log)
[2022-10-05_09h04m27s_Runtime.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9713616/2022-10-05_09h04m27s_Runtime.log)
[Event-Viewer_Error1.txt](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9713617/Event-Viewer_Error1.txt)
[Event-Viewer_Error2.txt](https://github.com/SafeExamBrowser/seb-win-refactoring/files/9713618/Event-Viewer_Error2.txt)

**Reply by dbuechel** [reply]
Thanks a lot, this indeed helps us greatly!

**Comment by yolpsoftware** [comment]
Thanks @AntonT76 for the error logs.

@dbuechel indeed the problem is that `document.body` is not yet set to a value, when the first Javascript invocation tries to get data from it. This is easy to fix by inserting these two lines on line 514 of Desktop/BrowserWindow.xaml.cs:

```
if (typeof __SEB_focusElement === 'undefined') {
  __SEB_focusElement = function (forward) {
    if (!document.body)                                                // <== INSERT THIS AND THE FOLLOWING LINE
      return;
    var items = [].map
....
```

Same for the "mobile" version.

If you tell me which branch, I can do the fix. Or do you want to do it yourself?

**Reply by dbuechel** [reply]
Great, I can add the line that's no problem. I'll also check whether the crash itself could be avoided by adding an access check in the event handler mentioned above. How were you able to reproduce the issue (so that I can test it myself)?

**Reply by yolpsoftware** [reply]
I have not reproduced the issue. But line 105 of the Client log AntonT76 provided is quite clear.

**Comment by dbuechel** [comment]
Created issue #480, please continue further discussion there. I'll add the changes proposed by @yolpsoftware and try to reproduce the issue (and possibly improve the error event handler) to be able to verify a fix.
